kani-driver/src/cbmc_property_renderer.rs

```diff
@@ -421,6 +421,103 @@ pub fn format_result(
+/// Loops through each of the check with a coverage property class and gives a status of FULL if all checks pertaining
+/// to a line number are SATISFIED. Similarly, it gives a status of NONE if all checks related to a line are UNSAT. If a line has both, it reports PARTIAL coverage.
+pub fn format_result_coverage(properties: &[Property]) -> String {
+    let mut formatted_output = String::new();
+    formatted_output.push_str("\nCoverage Results:\n");
+
+    let mut coverage_checks: Vec<&Property> =
+        properties.iter().filter(|&x| x.property_class() == "coverage").collect();
+
+    coverage_checks.sort_by_key(|check| (&check.source_location.file, &check.source_location.line));
+
+    let mut files: HashMap<String, Vec<(usize, CheckStatus)>> = HashMap::new();
+    for check in coverage_checks {
+        // Get line number and filename
+        let line_number: usize = check.source_location.line.as_ref().unwrap().parse().unwrap();
+        let file_name: String = check.source_location.file.as_ref().unwrap().to_string();
+
+        // Add to the files lookup map
+        files.entry(file_name.clone()).or_insert_with(Vec::new).push((line_number, check.status));
+    }
+
+    let mut coverage_results: HashMap<String, Vec<(usize, String)>> = HashMap::new();
+
+    // Loop through the files list and create a list of lines accessible in that file
+    for (file, val) in files {
+        let mut lines: HashSet<usize> = HashSet::new();
+        let mut line_results: Vec<(usize, String)> = Vec::new();
+        for check in val.clone() {
+            lines.insert(check.0);
+        }
+
+        // For each of these lines, create a map from line -> status
+        // example - {3 -> ["SAT", "UNSAT"], 4 -> ["UNSAT"] ...}
+        for line in lines.iter() {
+            let is_line_satisfied: Vec<_> = val
+                .iter()
+                .filter(|(line_number_accumulated, _)| *line == *line_number_accumulated)
+                .collect();
+
+            // Report lines as FULL if all of the coverage checks say SATISFIED, NONE if all of the coverage checks say UNSATISFIABLE,
+            // and PARTIAL if there is a mix of the two
+            let covered_status: String = if is_line_satisfied
+                .iter()
+                .all(|&is_satisfiable| is_satisfiable.1.to_string().contains("SATISFIED"))
+            {
+                "FULL".to_string()
+            } else if is_line_satisfied
+                .iter()
+                .all(|&is_satisfiable| is_satisfiable.1.to_string().contains("UNSATISFIABLE"))
+            {
+                "NONE".to_string()
+            } else {
+                "PARTIAL".to_string()
+            };
+
+            line_results.push((*line, covered_status));
+        }
+
+        line_results.sort_by_key(|&(line, _)| line);
+        coverage_results.insert(file.clone(), line_results);
+    }
```

Comment:
I was wondering if we could improve this a bit. I think you could actually process everything using one for loop.

Something like:

```rust
    #[derive(PartialEq, Eq, AsRefStr)]
    #[strum(serialize_all = "UPPERCASE")]
    enum CoverStatus {
        Full,
        Partial,
        None
    }
    let mut coverage_results : BTreeMap<String, BTreeMap<usize, CoverStatus>> = BTreeMap::default();
    for prop in properties {
        let src = prop.source_location;
        let file_entries = coverage_results.entry(src.file).or_default();
        let check_status = if prop.status == CheckStatus:: Satisfied { CoverStatus::Full } else { CoverStatus::None };
        file_entries.entry(src.line).and_modify(|line_status| {
            if line_status == check_status {
                line_status
            } else {
                CheckStatus::Partial
            }
        }.or_insert(check_status);
    }
```

Reply:
Done! Thanks for the suggestion